src/operator/contrib/multi_lamb-inl.h

```diff
@@ -0,0 +1,357 @@
+
+  bool all_inferred = true;
+  auto& input_shapes = *in_attrs;
+  auto& output_shapes = *out_attrs;
+
+  CHECK_EQ(param.learning_rates.ndim(), param.num_tensors)
+    << "Number of learning rates is inconsistent with num_tensors "
+    << "parameter passed. Expected number of learning rates: "
+    << param.num_tensors << ", and got " << param.learning_rates.ndim();
+  CHECK_EQ(param.wds.ndim(), param.num_tensors)
+    << "Number of weight decays is inconsistent with num_tensors "
+    << "parameter passed. Expected number of weight decays: "
+    << param.num_tensors << ", and got " << param.wds.ndim();
+  CHECK_EQ(param.step_count.ndim(), param.num_tensors)
+    << "Number of step counts is inconsistent with num_tensors."
+    << "Expected number of step counts: "
+    << param.num_tensors << ", and got " << param.step_count.ndim();
+
+  // Weights, gradients, mean and variance
+  for (int i = 0; i < param.num_tensors; ++i) {
+    mxnet::ShapeVector input_vec;
+    mxnet::ShapeVector output_vec({output_shapes[i]});
+    for (int j = 0; j < input_stride; ++j) {
+      input_vec.push_back(input_shapes[i * input_stride + j]);
+    }
+    all_inferred = all_inferred && ElemwiseShape<input_stride, 1>(attrs, &input_vec, &output_vec);
+  }
+  return all_inferred;
+}
+
+template <typename ParamType, int input_stride>
+inline bool MP_MultiLAMB_InferType(const nnvm::NodeAttrs& attrs,
+                                    std::vector<int> *in_attrs,
+                                    std::vector<int> *out_attrs) {
+  const ParamType& param = dmlc::get<ParamType>(attrs.parsed);
+  CHECK_EQ(in_attrs->size(), input_stride * param.num_tensors);
+  CHECK_EQ(out_attrs->size(), param.num_tensors);
+
+  bool all_inferred = true;
+  auto& input_types = *in_attrs;
+  auto& output_types = *out_attrs;
+
+  // weights, gradients
+  for (int i = 0; i < param.num_tensors; ++i) {
+    std::vector<int> input_vec;
+    std::vector<int> output_vec({output_types[i]});
+    for (int j = 0; j < 2; ++j) {
+      input_vec.push_back(input_types[i * input_stride + j]);
+    }
+    all_inferred = all_inferred &&
+            ElemwiseType<2, 1>(attrs, &input_vec, &output_vec);
+  }
+
+  // mean, var, weights32 (master copies of weights)
+  for (int i = 0; i < param.num_tensors; ++i) {
+    TYPE_ASSIGN_CHECK(input_types, input_stride * i + 2, mshadow::kFloat32);
+    TYPE_ASSIGN_CHECK(input_types, input_stride * i + 3, mshadow::kFloat32);
+    TYPE_ASSIGN_CHECK(input_types, input_stride * i + input_stride - 1, mshadow::kFloat32);
+  }
+  return all_inferred;
+}
+
+template<typename T>
+class LAMB_type_identity {
+ public:
+  using type = T;
+};
+
+template<typename T>
+class LAMB_single_precision {
+ public:
+  using type = float;
+};
+
+template<typename DType, typename MPDType>
+struct MultiLAMBKernelParam {
+  static const int N = 45;
+  size_t ntensors;
+  size_t max_size;
+  size_t total_size;
+  size_t sizes[N];
+  size_t tensor2temp_g[N];
```

Comment:
Can this op handle more than 45 tensors? Is it safe to sage MXNET_OPTIMIZER_AGGREGATION_SIZE > 45?